pkg/internal/server/grpc/validators.go

```diff
@@ -17,42 +17,102 @@ limitations under the License.
+		return "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSnaphotNameMissing)
+	}
+
+	snapshotHandle, driver, err := getVolSnapshotInfo(ctx, s.snapshotClient(), req.Namespace, req.SnapshotName)
+	if err != nil {
+		return "", status.Errorf(codes.Unavailable, msgUnavailableVolumeSnapshotNotReady)
 	}
 
-	return nil
+	if driver != s.driverName() {
+		return "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSnaphotDriverInvalidFmt, s.driverName())
+	}
+
+	return snapshotHandle, nil
 }
 
-func ValidateGetMetadataDeltaRequest(req *api.GetMetadataDeltaRequest) error {
-	// validate request
+func (s *Server) ValidateGetMetadataDeltaRequest(ctx context.Context, req *api.GetMetadataDeltaRequest) (string, string, error) {
 	if len(req.GetSecurityToken()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentSecurityTokenMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSecurityTokenMissing)
 	}
+
 	if len(req.GetNamespace()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentNamespaceMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentNamespaceMissing)
 	}
+
 	if len(req.GetBaseSnapshotName()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentBaseSnapshotNameMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentBaseSnapshotNameMissing)
 	}
+
 	if len(req.GetTargetSnapshotName()) == 0 {
-		return status.Errorf(codes.InvalidArgument, msgInvalidArgumentTargetSnapshotNameMissing)
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentTargetSnapshotNameMissing)
+	}
+
+	baseSnapshotHandle, driver, err := getVolSnapshotInfo(ctx, s.snapshotClient(), req.Namespace, req.BaseSnapshotName)
+	if err != nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableVolumeSnapshotNotReady)
 	}
 
-	return nil
+	if driver != s.driverName() {
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSnaphotDriverInvalidFmt, s.driverName())
+	}
+
+	targetSnapshotHandle, driver, err := getVolSnapshotInfo(ctx, s.snapshotClient(), req.Namespace, req.TargetSnapshotName)
+	if err != nil {
+		return "", "", status.Errorf(codes.Unavailable, msgUnavailableVolumeSnapshotNotReady)
+	}
+
+	if driver != s.driverName() {
+		return "", "", status.Errorf(codes.InvalidArgument, msgInvalidArgumentSnaphotDriverInvalidFmt, s.driverName())
+	}
+
+	return baseSnapshotHandle, targetSnapshotHandle, nil
+}
+
+// getVolSnapshotInfo returns snapshot handle and csi driver info of the VolumeSnapshot.
+func getVolSnapshotInfo(ctx context.Context, cli snapshot.Interface, namespace, vsName string) (string, string, error) {
+	vs, err := cli.SnapshotV1().VolumeSnapshots(namespace).Get(ctx, vsName, metav1.GetOptions{})
+	if err != nil {
+		return "", "", err
+	}
+	if vs.Status.ReadyToUse == nil || !*vs.Status.ReadyToUse {
+		return "", "", fmt.Errorf("VolumeSnapshot is not ready, name: %s", vsName)
```

Comment:
Use `status.Errorf` directly and not `fmt.Errorf`.   Don't use `status.Errorf` in the caller.